src/gennylib/src/topology.cpp

```diff
@@ -0,0 +1,135 @@
+// Copyright 2021-present MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string>
+#include <optional>
+#include <bsoncxx/json.hpp>
+#include <bsoncxx/types.hpp>
+#include <boost/log/trivial.hpp>
+
+#include <gennylib/topology.hpp>
+
+namespace genny {
+
+std::string Topology::nameToUri(const std::string& name) {
+    std::set<std::string> nameSet;
+    // Hostnames returned from some commands have the form configRepl/hostname:port
+    auto strippedName = name.substr(name.find('/') + 1, std::string::npos);
+    nameSet.insert(strippedName);
+    _factory.overrideHosts(nameSet);
+    return _factory.makeUri();
+}
+
+void Topology::getDataMemberConnectionStrings(DBService& service) {
+    auto res = service.runAdminCommand("isMaster");
+    if (!res.view()["setName"]) {
+        std::unique_ptr<MongodDescription> desc = std::make_unique<MongodDescription>();
+        desc->mongodUri = service.uri();
+        this->_topology.reset(desc.release());
+        return;
+    }
+
+    auto primary = res.view()["primary"];
+
+    std::unique_ptr<ReplSetDescription> desc = std::make_unique<ReplSetDescription>();
+    desc->primaryUri = nameToUri(std::string(primary.get_utf8().value));
+
+    auto hosts = res.view()["hosts"];
+    if (hosts && hosts.type() == bsoncxx::type::k_array) {
+        bsoncxx::array::view hosts_view = hosts.get_array();
+        for (auto member : hosts_view) {
+            MongodDescription memberDesc;
+            memberDesc.mongodUri = nameToUri(std::string(member.get_utf8().value));
+            desc->nodes.push_back(memberDesc);
+        }
+    }
+
+    // The "passives" field contains the list of unelectable (priority=0) secondaries
+    // and is omitted from the server's response when there are none.
+    auto passives = res.view()["passives"];
+    if (passives && passives.type() == bsoncxx::type::k_array) {
+        bsoncxx::array::view passives_view = passives.get_array();
+        for (auto member : passives_view) {
+            MongodDescription memberDesc;
+            memberDesc.mongodUri = std::string(member.get_utf8().value);
+            desc->nodes.push_back(memberDesc);
+        }
+    }
+
+    this->_topology.reset(desc.release());
```

Comment:
I think this would be more testable if you had a `MongodDescription` ctor that took in a `view_or_value` (which you can trivially construct without live connections to a db).

Reply:
I tried this, but unfortunately a bson `element` doesn't seem to be a `view`, so couldn't be passed in, unless I'm overlooking something. I tried doing two different constructors, one for `element` and one for `std::string`, but it turns out that the document `element` and the `array::element` are different. They share a private base class that I can't seem to figure out how to convince the compiler to let me `dynamic_cast` to.

I figured having three different implementations of that constructor might not be an improvement over the existing situation, but let me know if that'd still be better!